pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -457,9 +457,8 @@ func (p *PriorityQueue) movePodsToActiveQueue(pods []*v1.Pod) {
 
 // getUnschedulablePodsWithMatchingAffinityTerm returns unschedulable pods which have
 // any affinity term that matches "pod".
+// NOTE: this function assumes lock has been acquired in caller
 func (p *PriorityQueue) getUnschedulablePodsWithMatchingAffinityTerm(pod *v1.Pod) []*v1.Pod {
-	p.lock.RLock()
```

Comment:
@Huang-Wei - Does all the callers already acquire the lock ? In this case I'm fine with this.

Reply:
Right now there are two callers:

```go
func (p *PriorityQueue) AssignedPodAdded(pod *v1.Pod) {
	p.movePodsToActiveQueue(p.getUnschedulablePodsWithMatchingAffinityTerm(pod))
}

func (p *PriorityQueue) AssignedPodUpdated(pod *v1.Pod) {
	p.movePodsToActiveQueue(p.getUnschedulablePodsWithMatchingAffinityTerm(pod))
}
```

And in `p.movePodsToActiveQueue`, it's holding a `lock.Lock()`. I would say it's not an _exact_ case that callers holds the hock, especially if the logic got changed to:

```go
func (p *PriorityQueue) AssignedPod{Added,Updated}(pod *v1.Pod) {
	pods := p.getUnschedulablePodsWithMatchingAffinityTerm(pod)
	p.movePodsToActiveQueue(pods)
}
```

Reply:
To be honest, although the fix works in my env, personally I don't think it's that convincing. So I just put it here for more comments/brainstorming.

And according to my debug, I did see during the pod preemption, in some scheduling loop, lower priority pod even wasn't aware of that the pending high priority pod has been updated with "NominatedNode" info, and tricky thing is it's not always unaware of that - sometime it did, but sometime it didn't. So generally at this moment I'm suspecting it's very likely an issue of lock racing. 

```console
I1104 01:32:52.405448    3016 generic_scheduler.go:498] [DEBUG] podFitsOnNode for pod lab1-1-6bc98fd4b9-trvgj
I1104 01:32:52.405453    3016 generic_scheduler.go:429] [DEBUG] len(nominatedPods) = 1 <== it's good
I1104 01:32:52.435863    3016 generic_scheduler.go:498] [DEBUG] podFitsOnNode for pod lab1-1-6bc98fd4b9-trvgj
I1104 01:32:52.435869    3016 generic_scheduler.go:429] [DEBUG] len(nominatedPods) = 1 <== it's good
I1104 01:32:52.889055    3016 generic_scheduler.go:498] [DEBUG] podFitsOnNode for pod lab1-4-57fb9db7d7-2z6bg
I1104 01:32:52.889061    3016 generic_scheduler.go:429] [DEBUG] len(nominatedPods) = 0 <== it's good also
I1104 01:32:52.897087    3016 generic_scheduler.go:498] [DEBUG] podFitsOnNode for pod lab1-2-67ddd867d9-jpvww
I1104 01:32:52.897094    3016 generic_scheduler.go:429] [DEBUG] len(nominatedPods) = 0 <== it's not good
I1104 01:32:52.897199    3016 generic_scheduler.go:498] [DEBUG] podFitsOnNode for pod lab1-2-67ddd867d9-7mgwx
I1104 01:32:52.897203    3016 generic_scheduler.go:429] [DEBUG] len(nominatedPods) = 0 <== it's not good
I1104 01:32:52.897277    3016 generic_scheduler.go:498] [DEBUG] podFitsOnNode for pod lab1-2-67ddd867d9-bbqkn
I1104 01:32:52.897283    3016 generic_scheduler.go:429] [DEBUG] len(nominatedPods) = 1 <== it's good
I1104 01:32:52.905600    3016 generic_scheduler.go:498] [DEBUG] podFitsOnNode for pod lab1-2-67ddd867d9-bbqkn
```

Reply:
This function requires the lock. The main scheduler thread as well as event handlers can read/write `unschedulableQ` in different threads.

Reply:
@bsalamat yeap, still digging into the code to find out the real root cause.